### PR TITLE
Use ScopedThreadLocals for acquiring StringBuffers and Bytes.

### PR DESCRIPTION
**Pull request by nicktindall**
Depends on OpenHFT/Chronicle-Core#576

### CODE REVIEW

File: src/test/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueTest.java [hunk 1]
```diff
@@ -3571,7 +3573,7 @@ public void rollbackTest() {
                      builder(file, wireType).testBlockSize().build();
              ExcerptTailer tailer1 = queue.createTailer(named ? "named" : null)) {
 
-            StringBuilder sb = Wires.acquireStringBuilder();
+            StringBuilder sb = new StringBuilder();
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueTest.java:3576]: nice

File: src/test/java/net/openhft/chronicle/queue/impl/single/stress/RollCycleMultiThreadTest.java [hunk 2]
```diff
@@ -329,11 +330,13 @@ public synchronized Integer call() {
                 if (!dc.isPresent())
                     return -2;
 
-                StringBuilder sb = Wires.acquireStringBuilder();
-                dc.wire().read("say").text(sb);
+                try (ScopedResource<StringBuilder> stlSb = Wires.acquireStringBuilderScoped()) {
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/queue/impl/single/stress/RollCycleMultiThreadTest.java]: appreciate this is maybe going beyond the scope of this change, but maybe just have a StringBuilder field (from memory this class is per-thread?)

**Review by JerryShea — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/queue/impl/single/SCQIndexing.java [hunk 3]
```diff
@@ -194,18 +195,23 @@ private LongArrayValues arrayForAddress(@NotNull Wire wire, long secondaryAddres
...
+    private ValueIn readIndexValue(@NotNull WireIn w, @NotNull String expectedName) {
+        try (ScopedResource<StringBuilder> stlSb = Wires.acquireStringBuilderScoped()) {
+            final StringBuilder sb = stlSb.get();
+            long readPos = w.bytes().readPosition();
+            @NotNull final ValueIn valueIn = w.readEventName(sb);
+            if (!expectedName.contentEquals(sb))
+                throw new IllegalStateException("expecting " + expectedName + ", was " + sb + ", bytes: " + w.bytes().readPosition(readPos).toHexString());
+            return valueIn;
+        }
+    }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/single/SCQIndexing.java:213]: This pattern is repeated a few times, I wonder if `WireIn` should provide something like
```java
   public ValueIn expectEvent(String expectedName) throws UnexpectedValueException
```

File: src/main/java/net/openhft/chronicle/queue/impl/single/SCQTools.java [hunk 4]
```diff
@@ -66,12 +67,23 @@ private static MessageHistory readHistoryFromBytes(@NotNull final Wire wire, fin
...
+    private static ValueIn readHistoryValue(@NotNull final Wire wire) {
+        try (final ScopedResource<StringBuilder> stlSb = StoreTailer.SBP.acquireStringBuilderScoped()) {
+            StringBuilder sb = stlSb.get();
+            ValueIn valueIn = wire.read(sb);
+
+            if (!MethodReader.HISTORY.contentEquals(sb))
+                return null;
+            return valueIn;
+        }
+    }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/single/SCQTools.java:88]: similar thing again

File: src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java [hunk 5]
```diff
@@ -1116,6 +1113,18 @@ public SingleChronicleQueueStore acquire(int cycle, CreateStrategy createStrateg
...
+        private ValueIn readWireStoreValue(@NotNull Wire wire) throws StreamCorruptedException {
+            try (ScopedResource<StringBuilder> stlSb = Wires.acquireStringBuilderScoped()) {
+                StringBuilder name = stlSb.get();
+                ValueIn valueIn = wire.readEventName(name);
+                if (!StringUtils.isEqual(name, MetaDataKeys.header.name())) {
+                    throw new StreamCorruptedException("The first message should be the header, was " + name);
+                }
+                return valueIn;
+            }
+        }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java:1125]: and again

File: src/main/java/net/openhft/chronicle/queue/impl/table/SingleTableBuilder.java [hunk 6]
```diff
@@ -158,14 +159,20 @@ public TableStore<T> build() {
...
+    }
+
+    private ValueIn readTableStoreValue(@NotNull Wire wire) throws StreamCorruptedException {
+        try (ScopedResource<StringBuilder> stlSb = Wires.acquireStringBuilderScoped()) {
+            StringBuilder name = stlSb.get();
+            ValueIn valueIn = wire.readEventName(name);
+            if (!StringUtils.isEqual(name, MetaDataKeys.header.name())) {
+                throw new StreamCorruptedException("The first message should be the header, was " + name);
+            }
+            return valueIn;
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/table/SingleTableBuilder.java:175]: and again

File: src/main/java/net/openhft/chronicle/queue/impl/table/SingleTableStore.java [hunk 7]
```diff
@@ -295,11 +296,23 @@ public synchronized LongValue acquireValueFor(CharSequence key, final long defau
...
+    private static ValueIn readEventIfNameEquals(WireIn wireIn, CharSequence expected) {
+        try (ScopedResource<StringBuilder> stlSb = Wires.acquireStringBuilderScoped()) {
+            StringBuilder sb = stlSb.get();
+            final ValueIn valueIn = wireIn.readEventName(sb);
+            if (StringUtils.equalsCaseIgnore(expected, sb)) {
+                return valueIn;
+            }
+            return null;
+        }
+    }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/table/SingleTableStore.java:309]: And again

Comment (nicktindall) [reply src/main/java/net/openhft/chronicle/queue/impl/table/SingleTableStore.java:309]: But of course they all have slight variations, e.g.
- return null or throw exception when name doesn't match
- case sensitivity of the comparison
- the error type/message thrown

**Review by peter-lawrey — APPROVED** [review]

**Pull request merged**